library/kotlin/src/io/envoyproxy/envoymobile/RequestBuilder.kt

```diff
@@ -0,0 +1,159 @@
+  fun addBody(body: ByteArray?): RequestBuilder {
+    this.body = body
+    return this
+  }
+
+  /**
+   * Add a retry policy to use for this request.
+   *
+   * @param retryPolicy the {@link io.envoyproxy.envoymobile.RetryPolicy} for this request.
+   * @return this builder.
+   */
+  fun addRetryPolicy(retryPolicy: RetryPolicy?): RequestBuilder {
+    this.retryPolicy = retryPolicy
+    return this
+  }
+
+  /**
+   * Append a value to the header key.
+   *
+   * @param name the header key.
+   * @param value the value associated to the header key.
+   * @return this builder.
+   */
+  fun addHeader(name: String, value: String): RequestBuilder {
+    if (headers.containsKey(name)) {
+      headers[name]!!.add(value)
+    }
+
+    headers[name] = mutableListOf(value)
+    return this
+  }
+
+  /**
+   * Remove the value in the specified header.
+   *
+   * @param name the header key to remove.
+   * @param value the value to be removed.
+   * @return this builder.
+   */
+  fun removeHeader(name: String, value: String): RequestBuilder {
+    if (headers.containsKey(name)) {
+      headers[name]!!.remove(value)
+    }
+    return this
+  }
+
+  /**
+   * Remove all headers with this name.
+   *
+   * @param name the header key to remove.
+   * @return this builder.
+   */
+  fun removeHeaders(name: String): RequestBuilder {
+    headers.remove(name)
+    return this
+  }
+
+  /**
+   * Append a value to the trailer key.
+   *
+   * @param name the trailer key.
+   * @param value the value associated to the trailer key.
+   * @return this builder.
+   */
+  fun addTrailer(name: String, value: String): RequestBuilder {
+    if (trailers.containsKey(value)) {
+      trailers[name]!!.add(value)
+    }
+
+    trailers[name] = mutableListOf(value)
+    return this
+  }
+
+  /**
+   * Remove the value in the specified trailer.
+   *
+   * @param name the trailer key to remove.
+   * @param value the value to be removed.
+   * @return this builder.
+   */
+  fun removeTrailers(name: String): RequestBuilder {
+    trailers.remove(name)
+    return this
+  }
+
+  /**
+   * Remove the value in the specified trailer.
+   *
+   * @param name the trailer key to remove.
+   * @param value the value to be removed.
+   * @return this builder.
+   */
+  fun removeTrailer(name: String, value: String): RequestBuilder {
+    if (trailers.containsKey(name)) {
+      trailers[name]!!.remove(value)
+    }
+    return this
+  }
+
+  /**
+   * Creates the {@link io.envoyproxy.envoymobile.Request} object using the data set in the builder
+   *
+   * @return the {@link io.envoyproxy.envoymobile.Request} object
+   */
+  fun build(): Request {
+    return Request(
+        method,
+        url,
+        headers,
+        trailers,
+        body,
+        retryPolicy
+    )
+  }
+
+  internal fun setHeaders(headers: Map<String, List<String>>): RequestBuilder {
+    this.headers.clear()
+    headers.map { entry -> this.headers[entry.key] = entry.value.toMutableList() }
```

Comment:
Can we avoid side-effects within `map`? Same with `setTrailers`

Reply:
I think what you maybe really want here is `forEach`.

Reply:
Yep agreed